go/runtime/registry/host_rofl.go

```diff
@@ -0,0 +1,401 @@
+			Output:     rsp.Output,
+			Round:      rsp.Round,
+			BatchOrder: rsp.BatchOrder,
+			Proof:      proof,
+		}, nil
+	default:
+		// Just submit and forget.
+		err := rh.clientSrv.SubmitTxNoWait(ctx, submitRq)
+		if err != nil {
+			return nil, err
+		}
+		return &protocol.HostSubmitTxResponse{}, nil
+	}
+}
+
+func (rh *roflHostHandler) handleHostRegisterNotify(
+	ctx context.Context,
+	rq *protocol.HostRegisterNotifyRequest,
+) (*protocol.Empty, error) {
+	// Subscribe to event notifications.
+	if err := rh.eventNotifier.RegisterNotify(ctx, rq); err != nil {
+		return nil, err
+	}
+
+	return &protocol.Empty{}, nil
+}
+
+type roflAttachRuntimeCmd struct {
+	rt host.Runtime
+	ch chan<- error
+}
+
+type roflEventNotifierCmd struct {
+	// registerNotify is the command to register for notifications.
+	registerNotify *protocol.HostRegisterNotifyRequest
+	// attachRuntime is the command to attach a runtime host.
+	attachRuntime *roflAttachRuntimeCmd
+}
+
+type roflEventNotifier struct {
+	startOne cmnSync.One
+
+	runtime   Runtime
+	clientSrv runtimeClient.RuntimeClient
+	cmdCh     chan *roflEventNotifierCmd
+
+	logger *logging.Logger
+}
+
+func newROFLEventNotifier(runtime Runtime, clientSrv runtimeClient.RuntimeClient, logger *logging.Logger) *roflEventNotifier {
+	return &roflEventNotifier{
+		startOne:  cmnSync.NewOne(),
+		runtime:   runtime,
+		clientSrv: clientSrv,
+		cmdCh:     make(chan *roflEventNotifierCmd),
+		logger:    logger,
+	}
+}
+
+func (en *roflEventNotifier) start() {
+	en.startOne.TryStart(en.run)
+}
+
+func (en *roflEventNotifier) RegisterNotify(ctx context.Context, rq *protocol.HostRegisterNotifyRequest) error {
+	en.start() // Ensure event notifier is running.
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case en.cmdCh <- &roflEventNotifierCmd{registerNotify: rq}:
+	}
+	return nil
+}
+
+func (en *roflEventNotifier) AttachRuntime(rt host.Runtime) error {
+	en.start() // Ensure event notifier is running.
+
+	ch := make(chan error, 1)
+	en.cmdCh <- &roflEventNotifierCmd{attachRuntime: &roflAttachRuntimeCmd{rt, ch}}
+
+	select {
+	case <-time.After(roflAttachRuntimeTimeout):
+		return fmt.Errorf("timeout while attaching runtime")
+	case err := <-ch:
+		return err
+	}
+}
+
+func (en *roflEventNotifier) run(ctx context.Context) {
+	var (
+		rt    host.Runtime
+		blkCh <-chan *roothash.AnnotatedBlock
+
+		notifyBlocks bool
+		notifyTags   [][]byte
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case cmd := <-en.cmdCh:
+			// Process a command.
+			switch {
+			case cmd.registerNotify != nil:
+				// Update configuration.
+				notifyBlocks = cmd.registerNotify.RuntimeBlock
+
+				if re := cmd.registerNotify.RuntimeEvent; re != nil {
+					notifyTags = re.Tags
+				} else {
+					notifyTags = nil
+				}
+			case cmd.attachRuntime != nil:
+				// Attach runtime.
+				var err error
+				switch {
+				case rt != nil:
+					// Already attached.
+					err = fmt.Errorf("runtime already attached")
```

Comment:
Do we return this error when we start the load balancer?

Reply:
Actually I need to fix the load balancer to pass-through non-RONL as it doesn't make sense to load-balance those.

Reply:
Fixed.